src/main/java/redis/clients/jedis/ClientSideCache.java

```diff
@@ -1,71 +1,108 @@
 package redis.clients.jedis;
 
 import java.nio.ByteBuffer;
-import java.util.HashMap;
+import java.util.Collections;
+import java.util.HashSet;
 import java.util.List;
 import java.util.Map;
-
-import redis.clients.jedis.exceptions.JedisException;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.function.Function;
 import redis.clients.jedis.util.SafeEncoder;
 
-public class ClientSideCache {
+public abstract class ClientSideCache {
 
-  private final Map<ByteBuffer, Object> cache;
+  private final Map<ByteBuffer, Set<Long>> keyHashes;
+  private final ReentrantReadWriteLock rwl = new ReentrantReadWriteLock();
+  private final Lock readLock = rwl.readLock();
+  private final Lock writeLock = rwl.writeLock();
 
-  public ClientSideCache() {
-    this.cache = new HashMap<>();
+  protected ClientSideCache() {
+    this.keyHashes = new ConcurrentHashMap<>();
```

Comment:
It seems to me that we are only adding stuff to this map. Some more housekeeping will be needed, otherwise the map will only grow, we need to remove unused keys and stuff, at some point. Could this be in itself a `Cache`, either Caffeine, or Guava?

Reply:
Adding TODO comments for now.